### PR TITLE
1328 cherry pick bulk v2

### PR DESCRIPTION
**Pull request by jonahkaye**
Refs: #[1328](https://github.com/metriport/metriport-internal/issues/1328)

### Description
- responded to all of Rafs comments. Implemented all of them basically.
- deleted duplicated files from api that were in core and modified the relevant imports of deleted files.
- cherry picked commits from previous PR due to git history issues

### Testing

- Local
  - tested using local factory 
- Staging
  - branched to staging

### Release Plan

- [ ] no need to release NPM. Just one or two words docx change.
- [ ] Merge this
- [ ] release

### CODE REVIEW

**Review by jonahkaye — COMMENTED** [review]
I made all changes in the comments here: https://github.com/metriport/metriport/pull/1347

File: packages/core/src/domain/document-signing/document-bulk-signer-lambda.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      })
+      .promise()
+      .catch(error => {
+        console.log(
+          `Error invoking lambda ${this.lambdaName} with error ${JSON.stringify(
+            error
+          )}. The lambda name is likely wrong`
+        );
+      });
+  }
```
[the remark this answers is not in the document]
Comment (leite08) [reply packages/core/src/domain/document-signing/document-bulk-signer-lambda.ts:34]: Also, `JSON.stringify(error)` not always results  in the result we'd expect. We built `errorToString()` for that. But once we expose the promise upstream, error handling likely follows suit.

File: packages/core/src/util/capture.ts [hunk 2]
```diff
@@ -1,4 +1,77 @@
...
+  return Object.entries(captureContext.extra ?? {}).reduce(
+    (acc, [key, value]) => ({
+      ...acc,
+      [key]: typeof value === "string" ? value : stringify(value, null, 2),
+    }),
+    {}
+  );
+}
 
 export type Capture = {
```
Comment (leite08) [inline packages/core/src/util/capture.ts:76]: nit: Types on the top of the file please

**Review by leite08 — CHANGES_REQUESTED** [review]
In addition to the comments below, what's up with all those files being added/updated on the react-native-sdk package?

File: packages/utils/src/document/document-upload-post-process.ts
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/utils/src/document/document-upload-post-process.ts:1]: I use it to test pieces of the document uploading process. Not useless to me :)

File: packages/api/src/external/fhir/document/__tests__/get-documents.test.ts
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/api/src/external/fhir/document/__tests__/get-documents.test.ts:1]: Ok but then I have to move a dozen other files with it or create duplicates. I am of the opinion after including several files moves in my PRs, that moving files should be confined to their own entirely separate PRs. It crowds a PR with changes and makes them much harder to follow and review. This PR is already hard to follow with all the files being moved and imports being changed. 

I think we should start following the practice of having "moving PRs", and I am happy to make moving that test and everything else attached to it that needs to move as well the first "Moving PR"

File: packages/core/src/domain/bulk-get-document-url.ts [hunk 3]
```diff
@@ -1,5 +1,8 @@
-export const BulkGetDocUrlStatus = ["processing", "completed", "failed"] as const;
-export type BulkGetDocUrlStatus = (typeof BulkGetDocUrlStatus)[number];
+export enum BulkGetDocUrlStatus {
+  completed = "completed",
+  failed = "failed",
+  processing = "processing",
+}
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/core/src/domain/bulk-get-document-url.ts:5]: Ok thats fair. Will switch to literal types, but maybe in the future this is a good comprise. 
```
export const bulkGetDocUrlStatus = {
  processing: "processing",
  completed: "completed",
  failed: "failed",
} as const;

export type BulkGetDocUrlStatus = typeof bulkGetDocUrlStatus[keyof typeof bulkGetDocUrlStatus];
```
I can use the constant like I used the enum: bulkGetDocUrlStatus.completed which I think is more descriptive than just using the string "completed"

File: packages/core/src/external/fhir/document/get-documents.ts [hunk 4]
```diff
@@ -12,7 +12,7 @@ export async function getDocuments({
   documentIds = [],
 }: {
   cxId: string;
-  patientId: string;
+  patientId?: string;
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/core/src/external/fhir/document/get-documents.ts:15]: reverted

File: packages/core/src/domain/document-signing/document-bulk-signer-response.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import { DocumentReference } from "@metriport/api-sdk/medical/models/document";
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/core/src/domain/document-signing/document-bulk-signer-response.ts]: removed the dependency and the whole file since it was redundant

File: packages/lambdas/src/document-bulk-signer.ts [hunk 6]
```diff
@@ -14,15 +12,13 @@ const region = getEnvOrFail("AWS_REGION");
...
+  capture.setExtra({ lambdaName: lambdaName });
+  console.log(
+    `Running with envType: ${getEnvType()}, region: ${region}, ` +
+      `bucketName: ${bucketName}, cxId: ${cxId}, patientId: ${patientId}, requestId: ${requestId} `
+  );
 
-    await getSignedUrls(cxId, patientId, requestId, bucketName, region, apiURL);
-  }
-);
+  await signUrlsAndSendToApi(cxId, patientId, requestId, bucketName, region, apiURL);
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/lambdas/src/document-bulk-signer.ts:23]: I think that is confusing to use a local factory inside the lambda. The usage of DocumentDownloaderLocal in the lambda caused me a lot of confusion. I see the factories purpose as enabling the  api endpoint to both work during local development and during deployment in cloud env. 
Besides, we wouldnt be able to use the proper pattern of calling the makeDocumentBulkSigner since it would return a DocumentBulkSignerLambda instead of a DocumentBulkSignerLocal, so we wouldnt even be using the pattern properly. 
Unless you feel strongly about this I am going to leave this as is

File: packages/infra/lib/api-stack.ts [hunk 7]
```diff
@@ -1172,13 +1171,13 @@ export class APIStack extends Stack {
         SEARCH_ENDPOINT: searchEndpoint,
         SEARCH_INDEX: searchIndex,
         SEARCH_USERNAME: searchUserName,
-        SEARCH_PASSWORD: searchPassword,
+        SEARCH_SECRET_NAME: searchSecretName,
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/infra/lib/api-stack.ts:1174]: They are all used. The `makeSearchServiceQuery` function uses them in `core/src/external/opensearch/file-search-connector-factory.ts`, which is called a few functions down the call stack from `searchDocuments`

File: packages/infra/lib/api-stack.ts [hunk 8]
```diff
@@ -1172,13 +1171,13 @@ export class APIStack extends Stack {
...
         API_URL: `http://${apiService.loadBalancer.loadBalancerDnsName}`,
         ...(sentryDsn ? { SENTRY_DSN: sentryDsn } : {}),
       },
       layers: [lambdaLayers.shared],
-      memory: 512,
-      timeout: Duration.minutes(5),
+      memory: 1024,
+      timeout: Duration.minutes(15),
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/infra/lib/api-stack.ts:1180]: https://github.com/metriport/metriport/pull/1252#discussion_r1414360075
https://github.com/metriport/metriport/pull/1252#discussion_r1414361036

File: packages/api/src/external/aws/document-bulk-signer-factory.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { DocumentBulkSignerLambda } from "@metriport/core/domain/document-signing/document-bulk-signer-lambda";
+import { DocumentBulkSignerLocal } from "@metriport/core/domain/document-signing/document-bulk-signer-local";
+import { DocumentBulkSigner } from "@metriport/core/domain/document-signing/document-bulk-signer";
+import { Config } from "@metriport/core/util/config";
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/api/src/external/aws/document-bulk-signer-factory.ts:4]: api using its own version of config

File: packages/infra/lib/api-stack.ts [hunk 10]
```diff
@@ -362,6 +362,21 @@ export class APIStack extends Stack {
...
+      medicalDocumentsBucket: medicalDocumentsBucket,
+      fhirServerUrl: props.config.fhirServerUrl,
+      envType: props.config.environmentType,
+      sentryDsn: props.config.lambdasSentryDSN,
+      alarmAction: slackNotification?.alarmAction,
+      searchEndpoint: ccdaSearchDomain.domainEndpoint,
+      searchIndex: ccdaSearchIndexName,
+      searchUserName: ccdaSearchUserName,
+      searchSecretName: ccdaSearchSecret.secretName,
+      loadBalancerDnsName: props.config.loadBalancerDnsName,
```
Comment (jonahkaye) [inline packages/infra/lib/api-stack.ts:377]: needs load balancer address but lambda is created before API Service so need to use the loadBalancer address in the config

File: packages/core/src/external/opensearch/file-search-connector-factory.ts [hunk 11]
```diff
@@ -1,16 +1,18 @@
+import { getSecret } from "@aws-lambda-powertools/parameters/secrets";
 import { OpenSearchFileIngestor } from "./file-ingestor";
 import { OpenSearchFileIngestorDirect } from "./file-ingestor-direct";
 import { OpenSearchFileIngestorSQS } from "./file-ingestor-sqs";
 import { OpenSearchFileSearcher } from "./file-searcher";
 import { OpenSearchFileSearcherDirect } from "./file-searcher-direct";
 import { Config } from "../../util/config";
 
-export function makeSearchServiceIngest(): OpenSearchFileIngestor {
+export async function makeSearchServiceIngest(): Promise<OpenSearchFileIngestor> {
```
Comment (jonahkaye) [inline packages/core/src/external/opensearch/file-search-connector-factory.ts:9]: had to make these async because getSecret is async: https://aws.amazon.com/blogs/compute/retrieving-parameters-and-secrets-with-powertools-for-aws-lambda-typescript/

File: packages/core/src/external/fhir/document/search-documents.ts [hunk 12]
```diff
@@ -65,7 +65,7 @@ async function searchOnCCDAFiles(
   contentFilter?: string
 ): Promise<DocumentReference[]> {
   if (!contentFilter) return [];
-  const searchService = makeSearchServiceQuery();
+  const searchService = await makeSearchServiceQuery();
```
Comment (jonahkaye) [inline packages/core/src/external/fhir/document/search-documents.ts:68]: ditto on getSecret being async requiring this to be awaited

File: packages/api/src/shared/config.ts [hunk 13]
```diff
@@ -268,9 +268,7 @@ export class Config {
   static getSearchUsername(): string {
     return getEnvVarOrFail("SEARCH_USERNAME");
   }
-  static getSearchPassword(): string {
```
Comment (jonahkaye) [inline packages/api/src/shared/config.ts:271]: passwords are secrets not env vars. see getSecret()

File: packages/core/src/domain/document-bulk-signer-response.ts [hunk 14]
```diff
@@ -1,14 +1,14 @@
 import { z } from "zod";
 
-export type DocumentBulkSignerLambdaResponse = {
+export type DocumentReferenceWithURL = {
```
Comment (jonahkaye) [inline packages/core/src/domain/document-bulk-signer-response.ts:3]: changed type name to be more descriptive

File: packages/infra/config/env-config.ts [hunk 15]
```diff
@@ -24,7 +24,7 @@ export type EnvConfig = {
   authSubdomain: string; // Authentication subdomain
   dbName: string;
   dbUsername: string;
-  loadBalancerDnsName?: string;
+  loadBalancerDnsName: string;
```
Comment (jonahkaye) [inline packages/infra/config/env-config.ts]: optional -> required since Bulk signing Lambda now needs it

File: packages/api/src/external/fhir/document/__tests__/get-documents.test.ts
[the remark this answers is not in the document]
Comment (leite08) [reply packages/api/src/external/fhir/document/__tests__/get-documents.test.ts:1]: I agree that we benefit from smaller PRs, 💯 

But its better to have changes that are similar by context than cross section PRs, touching multiple areas but not completing them.

In this case, it would be better to move the `get-documents` and its test in one PR, and have another PR to completely update another file/usecase, than moving all logic in one and all tests in another. It's hard to keep track of things and if we only merge one PR we don't complete the work to be done (which include the file and its dependencies).

If you want to create another PR as downstream of this one to keep things sane, glad to review them like so.

File: packages/utils/src/document/document-upload-post-process.ts
[the remark this answers is not in the document]
Comment (leite08) [reply packages/utils/src/document/document-upload-post-process.ts:1]: Sounds good, just wanted to make sure it's being used. Can you add a note, as we've been keeping this pattern of having a readme/doc at the top of scripts, please? Explaining what it does (why it exists) and how to use it.

File: packages/core/src/domain/bulk-get-document-url.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (leite08) [reply packages/core/src/domain/bulk-get-document-url.ts]: While your suggestion works just like an Enum, its quite complex. 

As for "more descriptive":
- we usually reference the value/option in a context, like `setStatus("completed")` - this is already very descriptive and having to type `bulkGetDocUrlStatus` just takes longer and makes the code more cluttered.

If you like more context when reading code, could consider installing an extension like [this one](https://marketplace.visualstudio.com/items?itemName=liamhammett.inline-parameters), to display inline param names.

It seems a good topic to a TechTalk.

File: packages/core/src/external/opensearch/search-documents.ts
[the remark this answers is not in the document]
Comment (leite08) [reply packages/core/src/external/opensearch/search-documents.ts:1]: Sounds good 👍

File: packages/api/src/shared/config.ts (hunk 13 above)
Comment (leite08) [reply packages/api/src/shared/config.ts:271]: The API does get secrets as env vars, see [api-service.ts](https://github.com/metriport/metriport/blob/5e3dea257861d8545b5edc8f8cc990a1c3426fbf/packages/infra/lib/api-stack/api-service.ts#L124).

If we're not using getting it from vars any longer, we could stop sending it over.

File: packages/core/src/domain/document-signing/bulk-sign.ts [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+import { DocumentReference } from "@medplum/fhirtypes";
+import { capture } from "../../util/capture";
+import { S3Utils } from "../../external/aws/s3";
+import { DocumentReferenceWithURL } from "../document-bulk-signer-response";
+import { searchDocuments } from "../../external/fhir/document/search-documents";
+import { MetriportError } from "../../util/error/metriport-error";
+import { BulkGetDocUrlStatus } from "../bulk-get-document-url";
+
+const ossApi = axios.create();
+dayjs.extend(duration);
```
Comment (leite08) [inline packages/core/src/domain/document-signing/bulk-sign.ts:13]: nit: I suggest we keep calls to `dayjs.extend()` before any const so we avoid using the extension before it gets initialized.

File: packages/lambdas/src/document-bulk-signer.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (leite08) [reply packages/lambdas/src/document-bulk-signer.ts:23]: I didn't say anything about using a factory. 😄 

I just suggested to use the exact same code the API uses when being run local, this way we prevent any updates on `DocumentBulkSignerLocal` that would result in diff behavior when testing local vs. running it on the cloud.

Maybe what's confusing is the usage of "Local" as the implementation that directly calls the service instead of going through a lambda. I've thought about that too, maybe a better name would be "Direct"?

Lastly, I didn't understand the point about `makeDocumentBulkSigner`, I don't see it being used in neither of the files.

Happy to huddle.

File: packages/infra/config/env-config.ts (hunk 15 above)
Comment (leite08) [reply packages/infra/config/env-config.ts]: It was optional b/c the first time the API gets provisioned this is not known. There's a cyclical dependency there, that sucks - ideally, we would use pre-defined internal URLs, like `api.staging.metriport.com` or `api.internal.staging.metriport.com`, but we didn't do this in the beginning.

Spinning off a new env would look like this:
- deploy API Stack (that lambda doesn't get created);
- update env config with the LB address;
- deploy again (lambda gets created).

If we make this required, we can't deploy the API Stack on new envs.
